[Blog] Batch Processing Semantic Highlighting in OpenSearch 3.3 (#3978)
opensearch-project/project-website: _posts/2025-10-13-batch-processing-semantic-highlighting.md

---
layout: post
title: "Batch processing semantic highlighting in OpenSearch 3.3"
authors:
 - junqiu
date: 2025-10-13
categories:
 - technical-post
meta_keywords: "semantic highlighting, batch processing, OpenSearch 3.3, ML inference, remote models, performance optimization"

> meta_description: "OpenSearch 3.3 introduces batch processing for remote semantic highlighting models, reducing ML inference calls and delivering 100--1,300% performance improvements."

Reviewer: ```suggestion
meta_description: "Learn how OpenSearch 3.3's new batch processing for semantic highlighting delivers up to 1,300% performance improvement through reduced ML inference calls and optimized GPU utilization."
```

> excerpt: "OpenSearch 3.3 introduces batch processing for remote semantic highlighting models, reducing ML inference calls from N to 1 per search query. Our benchmarks demonstrate 100-1300% performance improvements depending on document length and result set size."

Reviewer 2: ```suggestion
excerpt: "OpenSearch 3.3 introduces batch processing for remote semantic highlighting models, reducing ML inference calls from N to 1 per search query. Our benchmarks demonstrate 100--1,300% performance improvements, depending on document length and result set size."
```

has_science_table: true
---

In OpenSearch 3.0, we introduced [semantic highlighting](https://opensearch.org/blog/introducing-semantic-highlighting-in-opensearch/)---an AI-powered feature that intelligently identifies relevant passages in search results based on meaning rather than exact keyword matches.

OpenSearch 3.3 introduces batch processing for externally hosted semantic highlighting models, reducing machine learning (ML) inference calls from N to 1 per search query. Our benchmarks demonstrate 100--1300% performance improvements depending on document length and result set size.

> **Try our demo now:** Experience semantic highlighting on the [OpenSearch ML Playground](https://ml.playground.opensearch.org/), presented in the following image.

Reviewer 2: ```suggestion
**Try our demo now**: Experience semantic highlighting on the [OpenSearch ML Playground](https://ml.playground.opensearch.org/), presented in the following image.
```

| k value | Search client | Document length | P50 without batch processing (ms) | P50 with batch processing (ms) | P50 improvement | P90 without batch processing (ms) | P90 with batch processing (ms) | P90 improvement |
|---|---|---|---|---|---|---|---|---|
| 10 | 1 | Long | 209 | 123 | 70% | 262 | 179 | 46% |
| 10 | 4 | Long | 378 | 171 | 121% | 487 | 302 | 61% |
| 10 | 8 | Long | 699 | 309 | 126% | 955 | 624 | 53% |
| 10 | 1 | Short | 175 | 55 | 218% | 217 | 59 | 268% |
| 10 | 4 | Short | 327 | 62 | 427% | 445 | 120 | 271% |
| 10 | 8 | Short | 610 | 101 | 504% | 860 | 227 | 279% |
| 50 | 1 | Long | 867 | 633 | 37% | 999 | 717 | 39% |
| 50 | 4 | Long | 1,937 | 912 | 112% | 2,248 | 1,685 | 33% |
| 50 | 8 | Long | 3,638 | 1,474 | 147% | 4,355 | 3,107 | 40% |
| 50 | 1 | Short | 760 | 82 | 827% | 828 | 205 | 304% |
| 50 | 4 | Short | 1,666 | 193 | 763% | 1,971 | 362 | 445% |

> | 50 | 8 | Short | 3,162 | 219 | 1344% | 3,704 | 729 | 408% |

Reviewer 2: ```suggestion
| 50 | 8 | Short | 3,162 | 219 | 1,344% | 3,704 | 729 | 408% |
```

The benchmarking demonstrates that batch processing reduces the semantic highlighting overhead. For short documents with k=50 and 8 clients, batch processing reduces highlighting latency from 3,162ms to just 219 ms (P50)---a **1344% improvement**. The P90 latency also shows improvements (408%), demonstrating consistent performance benefits. The semantic search baseline (~25 ms) remains constant, so these improvements directly translate to faster end-to-end response times.

**Key findings:**

* **k=10**: Moderate to significant improvement (46-504% for P50, 46-279% for P90).

> * **k=50**: Dramatic improvement (37-1344% for P50, 33-445% for P90).

Reviewer 2: ```suggestion
* **k=50**: Dramatic improvement (37--1,344% for P50, 33--445% for P90).
```

* **Short documents benefit more**: Up to 1344% faster (P50) vs 147% for long documents at k=50.

>   * **Why the difference?** Long documents could exceed the model's 512 token limit, requiring multiple chunked inference runs even with batch processing. Short documents can be processed in a single pass, maximizing the benefit of batching.

Reviewer 2: ```suggestion
  * **Why the difference?**: Long documents could exceed the model's 512 token limit, requiring multiple chunked inference runs even with batch processing. Short documents can be processed in a single pass, maximizing the benefit of batching.
```

| K-value | Search clients | Doc length | Without batch (ops/s) | With batch (ops/s) | Improvement |
|---|---|---|---|---|---|
| 10 | 1 | Long | 4.23 | 6.29 | 49% |
| 10 | 4 | Long | 9.18 | 17.9 | 95% |
| 10 | 8 | Long | 10.02 | 21.27 | 112% |
| 10 | 1 | Short | 4.83 | 11.59 | 140% |
| 10 | 4 | Short | 10.47 | 37.79 | 261% |
| 10 | 8 | Short | 12.03 | 48.33 | 302% |
| 50 | 1 | Long | 1.11 | 1.49 | 34% |
| 50 | 4 | Long | 1.99 | 3.74 | 88% |
| 50 | 8 | Long | 2.12 | 4.28 | 102% |
| 50 | 1 | Short | 1.27 | 4.3 | 239% |
| 50 | 4 | Short | 2.27 | 11.55 | 409% |
| 50 | 8 | Short | 2.43 | 14.33 | 490% | |

Throughput improvements demonstrate that batch processing not only reduces individual query latency but also increases the overall system capacity, allowing you to serve more concurrent users with the same infrastructure.

## Conclusion

Batch processing in OpenSearch 3.3 brings significant performance improvements to semantic highlighting for remote models. By reducing the number of ML inference calls from N to 1 per search, we've delivered:

* Faster response times and higher query throughput when highlighting multiple search results
* More efficient use of remote model resources

> * Backward compatible with existing queries work unchanged

Reviewer 2: ```suggestion
* Backward-compatible queries (existing queries work as is).
```

authors:
 - junqiu
date: 2025-10-13
categories:
 - technical-post
meta_keywords: "semantic highlighting, batch processing, OpenSearch 3.3, ML inference, remote models, performance optimization"
meta_description: "OpenSearch 3.3 introduces batch processing for remote semantic highlighting models, reducing ML inference calls and delivering 2-14× performance improvements."
excerpt: "OpenSearch 3.3 introduces batch processing for remote semantic highlighting models, reducing ML inference calls from N to 1 per search query. Our benchmarks demonstrate 2-14× performance improvements depending on document length and result set size."
has_science_table: true
---

We introduced [semantic highlighting](https://opensearch.org/blog/introducing-semantic-highlighting-in-opensearch/) in OpenSearch 3.0—an AI-powered feature that intelligently identifies relevant passages in search results based on meaning rather than exact keyword matches.

OpenSearch 3.3 introduces batch processing for remote semantic highlighting models, reducing ML inference calls from N to 1 per search query. Our benchmarks demonstrate 2-14× performance improvements depending on document length and result set size.

> **Try it now:** Experience batch semantic highlighting on the [OpenSearch ML Playground](https://ml.playground.opensearch.org/) [TODO: Add direct link to batch semantic highlighting demo].

Author: Feature onboarded on ML playground
